Albedo decay based on burn mask in `topo.nc`

jomey opened the pull request:
Closes #26 

**NOTE**
Once merged  - all configuration `.ini` files will need to update their parameter for time decay window.

## Renamed parameters in `[albedo]` section
```
date_method_start_decay -> decay_start
date_method_end_decay -> decay_end
```

jomey committed:
Envphys - Lint albedo code; fix typos

jomey committed:
Albedo - Rename decay start and end date config and break out method.

Rename the parameters to define a decay window and move the function to determine whether
the window is present to the distribute class for re-use. This prepares add a new post
fire based decay method that should also be triggered within the same time window.

The config parameters are now:
* start_decay (formerly date_method_start_decay)
* end_decay (formerly date_method_end_decay)

jomey committed:
Topo - Read in a burn mask if present.

Adds reading in a new layer that defines a burn mask on the topo file.

jomey committed:
Distribute - Albedo - Lint code

jomey committed:
Albedo - Add new decay function based on binary burn mask

Add a new option to decay albedo based on a burn mask in the topo file and configured
values in the .ini via `post_fire_k_burned` and `post_fire_k_unburned`. The method itself
will be triggered via the decay method `post_fire`.

Co-authored-by: caldwellng <caldwellng@users.noreply.github.com>

jomey committed:
Precip - Use failsafe access to config key for storm_days_restart

jomey committed:
Envphys - Small tweaks to decay_burned function

* Fix type hint for `last_snow` parameter; it's an array
* Improve output initialization and use empty values
* Rename exponential variables that conflicted with method name

jomey committed:
CoreConfig - Albedo - Change the name for dates one more time

Change the names to configure dates to `decay_start` and `decay_end` to be consistent with
the main parameter name `decay_method`.

jomey committed:
Tests - Distribute Albedo - Use proper data type for fake cos_z

jomey committed:
Distribute - Albedo - Rename variables to be PEP 8 conform and readable

jomey committed:
Distribute - Albedo - Follow up fixes after merge with latest in main

The main branch had an overhaul of the init method, changing a few pieces of the logic in
this branch with was developed in parallel. This now also adds the accessor to veg_type
from the topo with the ImageData base class.

jomey committed:
Copilot review - Remove duplicat code; fix description strings

jomey committed:
Tests - Fix after merge with latest in main.

Setup changed when testing distribution variables.

jomey committed:
Envphys - Albedo - Fix decay rate for default values

Bug introduced with the refactor to add burned albedo and used the current passed time
versus the decay window hours. Thank you Copilot!
Added tests to catch this in future.

jomey committed:
Enphys - Albedo - Use numexpr for decay_burned function

jomey committed:
Albedo - Fix latest merge with main

Couple of conflicts got commited unresolved. This mostly addresses the test class for the
distribute/albedo. Also restores some logic improvement when deciding on the distribute
method to use.

jomey committed:
Envphys - Albedo - Simplify the power decay albedo logic

Reduce the complexity of the method by combining common steps between conditional steps.
This eliminates all but one if case.

jomey committed:
Albedo - Add debug logging for chosen method and add error handling for post_fire

Added a check that k burned and unburned are set when chossing this method.

jomey committed:
Albedo - Post Fire - Only use a different decay for burned pixels.

Change of strategy and update the logic to apply standard time-decay albedo for unburned
pixels and only a different decay for burned ones. This also uses the initial "clean"
albedo after a storm for burned pixels for the burned pixels and then applies the burn
decay rate when there is no more new snow.

jomey committed:
Envphys - Albedo - Ensure that all constants are float32

Numexpr will throw an error when trying to write float64 results back into a float32.

Also partially addresses issue in #59

jomey committed:
Albedo - Update argument and return types using numpy's typing module

jomey committed:
Tests - Topo - Fix faulty merge with main

jomey committed:
Envphys - albedo - burn decay - Start decay once one day has passed

With the `exp` function using the last snow as a multiplier, any period less than a day
caused an increase an albedo before. To prevent these short-term spikes, the burn decay is
applied after one full day sine a storm.

jomey committed:
Distribute - Albedo - Change threshold for the burned and no snowfall mask

Changing the threshold for the burn mask to be active to be larger than one full day.
At a storm day value of 1.0, the time decay method a decay rate of 0.09817055, versus the
exponential burned rate of 0.94176453. This causes a jump and increase of the previous
albedo.

jomey committed:
Tests - Distribute - Add more attributes to the Topo mock

That is DEM and vegetation type

jomey committed:
Albedo - Post Fire - Change fresh snowfall logic one more time.

Third times the charm and this changes the logic of a post fire configuration right after
fresh snow. Previously, to catch the lower albedo values late in the melt season after a
"reset" event, the post fire decay was only applied after one day since snowfall. This,
however, created edge cases where the albedo jumped up right after the one-day mark when
the post fire decay rate still was slower than the time decay. To alleviate the need to
find a minimum time since last snowfall, this logic now always applies time decay and then
the post fire decay for pixels in the burn mask and that are lower than the time decay
value.

brentwilder commented on smrf/envphys/albedo.py:
```diff
@@ -1,17 +1,19 @@
-IRZ0 = 0.1              # ir zenith increase range, gsize=0
-BOIL = 373.15           # boiling temperature K
-GRAVITY = 9.80665       # gravity (m/s^2)
+MAXV = 1.0  # vis albedo when gsize = 0
```
Is there any particular reason these not just imported from `constants.py`? Could be easier to manage one list and just import where needed.

brentwilder commented on smrf/envphys/albedo.py:
```diff
@@ -239,24 +210,81 @@ def decay_alb_hardy(litter, veg_type, storm_day, alb_v, alb_ir):
+    Returns: Tuple
+        alb_v_d, alb_ir_d : numpy arrays of decayed albedo
+    """
+    if k_burned is None:
```
Optional, but I would raise this error one layer higher into `distribute/albedo.py`. 

In my eyes it looks ike `envphys/albedo.py` is handling most of the math, it would be nice for config checks/ settings to be handled by `distribute/albedo.py`. 

Same result, but may be cleaner imo.

brentwilder commented on smrf/framework/CoreConfig.ini:
```diff
@@ -866,6 +864,19 @@ description = Full path to netcdf files that contain albedo values.
               Once set none of the other albedo variables will be effective and the
               raw values from the files will be used.
 
+post_fire :
+default = False,
+type = bool,
+description = Apply a different post fire decay method for areas masked in the topo
+              file via a 'burn_mask' layer. This also requires to set the
+              'post_fire_k_burned' value in this configuration section
+
+post_fire_k_burned:
+default = None,
+type = float,
+description = Decay power for areas masked as 1 (burned) in the burned mask of the
+              topo file
```
It's common in the fire literature to see data products like low, med, high soil-burn-severity. Would this be easy to implement in the future if a student/researcher was interested in having different parameterizations for varying burn severities? There is some evidence that it matters (Uecker et al., 2020), which is why I poke on this. I think this is something more than this PR, but I think just doing a pass to see if it would be doable for determined student to figure this out. **Or if there any small abstractions that could be made to enable this a little easier in future.**

jomey replied:
I like the idea of potentially expanding this to use different values based on burn severity. As you said though, this is well beyond the scope of this research and PR. 

Thinking out loud, an easy extension would be to refactor this and instead of using a simple binary mask and define the k values as the mask and then do a per pixel calculation instead of one constant for the entire area. Think this could be achieved with not much code changes and even simplify the logic.

In the end, this is a great foundation to do more post fire albedo work.

brentwilder replied:
Perfect yeah this was more the logic check i was thinking to ensure it works cleanly for future extensions.

brentwilder commented on smrf/framework/CoreConfig.ini:
```diff
@@ -866,6 +864,19 @@ description = Full path to netcdf files that contain albedo values.
+              'post_fire_k_burned' value in this configuration section
+
+post_fire_k_burned:
+default = None,
```
I think default should be None, but also think it would be nice to know what values to use for `post_fire_k_burned`? If there is general guidance based on the paper perhaps this could be added to the description above with a citation?

Just thinking that may make it easier for others to pick up this model / know where to look for appropriate decay value.

jomey replied:
Think we have a :baby_chick: and :egg: problem. We need this to merge to make it part of the paper and create a citation we can add once published. Added #74

brentwilder replied:
Got it! No that makes sense, thanks for creating an issue to note it.

brentwilder reviewed (approved):
I think this is a great addition to the community model and only have very minor comments to enable others to use the new functionality a little bit easier.

jomey committed:
Envphys - Constants - Re-use already defined values for albedo

Great catch by @brentwilder that the albedo module had the same constants defined as the
global constant file. Used this chance to rename them to (hopefully) be a bit more
readable.

Also organized the constants file and group values with sorting by alphabetically order.

jomey committed:
Albedo - Move post_fire_k_burned check to distribute modules

Great feedback by @brentwilder that math should be done in the envphys module while
distribute checks for presence of all relevant parameters.

jomey committed:
CoreConfig - Add TODO to post_fire_k_burned parameter

jomey committed:
Distribute - Albedo - Improve docstring describing the date_method
